### PR TITLE
RavenDB-20416: fix updating subscription with LastDocument if last doc is from replication

### PR DESCRIPTION
**Pull request by garayx**
https://github.com/ravendb/ravendb/pull/16398

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Collections/CollectionsHandlerProcessorForGetLastChangeVector.cs [hunk 1]
```diff
@@ -22,7 +22,12 @@ protected override async ValueTask HandleCurrentNodeAsync()
             using (ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
             using (context.OpenReadTransaction())
             {
-                var result = RequestHandler.Database.DocumentsStorage.GetLastDocumentChangeVector(context.Transaction.InnerTransaction, context, collection);
+                long lastEtag = collection == Client.Constants.Documents.Collections.AllDocumentsCollection
+                    ? DocumentsStorage.ReadLastDocumentEtag(context.Transaction.InnerTransaction)
+                    : RequestHandler.Database.DocumentsStorage.GetLastDocumentEtag(context.Transaction.InnerTransaction, collection);
+
+                var result = RequestHandler.Database.DocumentsStorage.GetNewChangeVector(context, lastEtag);
+
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Processors/Collections/CollectionsHandlerProcessorForGetLastChangeVector.cs]: `GetLastDocumentChangeVectorForSubscription` ?

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 2]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
+                                        return null;
+                                    }
+
+                                    return null;
+                                }
+
+                                return null;
+                            }
+
+                            return null;
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB_20416.cs]: can we have a nicer function for this?

**Review by karmeli87 — APPROVED** [review]

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 3]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
-                    var connectionState = db2.SubscriptionStorage.GetSubscriptionConnectionsState(ctx, state.SubscriptionName);
-                    if (connectionState != null)
+                    var newCv = cvs[db.Name];
+                    var cv = WaitForValue(() =>
+                    {
+                        using (db.ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
+                        using (ctx.OpenReadTransaction())
+                        {
+                            var connectionState = db.SubscriptionStorage.GetSubscriptionConnectionsState(ctx, state.SubscriptionName);
+                            if (connectionState != null)
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_20416.cs]: reduce the nesting here

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 4]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
+                            {
+                                var connection = connectionState.GetConnections().FirstOrDefault();
+                                if (connection != null)
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_20416.cs]: reduce the nesting here

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 5]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
+                                {
+                                    var subsState = connection.SubscriptionState;
+                                    if (subsState.ShardingState != null)
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_20416.cs]: reduce the nesting here

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 6]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
+
+                            return null;
+                        }
+                    }, newCv, interval: 333);
+                    Assert.Equal(newCv, cv);
+
+                    DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal,
+                        "enable the commented code after RavenDB-20454 is done");
+
+                    /*var cvList = cv.ToChangeVectorList();
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_20416.cs]: clean it up please

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 7]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
-                using (db2.ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
-                using (ctx.OpenReadTransaction())
+                var shards1 = Sharding.GetShardsDocumentDatabaseInstancesFor(store1);
+                Assert.NotNull(shards1);
+                var dbIds1 = await shards1.Select(x => x.DbBase64Id).ToListAsync();
+
+                var shards2 = Sharding.GetShardsDocumentDatabaseInstancesFor(store2);
+                Assert.NotNull(shards2);
+                var dbIds2 = await shards2.Select(x => x.DbBase64Id).ToListAsync();
+                Dictionary<string, string> cvs2 = new Dictionary<string, string>();
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_20416.cs:222]: check the warnings in this PR

File: test/SlowTests/Issues/RavenDB_20416.cs [hunk 8]
```diff
@@ -167,37 +204,120 @@ private async Task CheckSubscriptionFetchAfterUpdate(DocumentStore store1, IDocu
...
+                var db1 = await Databases.GetDocumentDatabaseInstanceFor(store1);
+                var dbId1 = db1.DbBase64Id;
+                var db2 = await Databases.GetDocumentDatabaseInstanceFor(store2);
+                var dbId2 = db2.DbBase64Id;
 
-            Assert.Equal(2, cvList.Count);
-            Assert.Equal(count, cvList.FirstOrDefault(x => x.DbId == dbId1).Etag);
-            Assert.Equal(count + 1, cvList.FirstOrDefault(x => x.DbId == dbId2).Etag);
+                List<ChangeVectorEntry> cvList = null;
+                var tt = await WaitForValueAsync(() =>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_20416.cs]: why async?

File: test/SlowTests/Issues/RavenDB_20416.cs (hunk 6 above)
Comment (garayx) [reply test/SlowTests/Issues/RavenDB_20416.cs]: this needs to be uncommented after  RavenDB-20454 is done

**Pull request merged**